tests/MongoDB.Driver.Tests/ClientSessionHandleTests.cs

```diff
@@ -258,45 +278,489 @@ public void StartTransaction_should_call_coreSession()
+
+        private Mock<IClock> CreateClockMock(DateTime now, bool[] isRetryAttemptsWithTimeout, bool shouldNowBeAdded)
+        {
+            if (isRetryAttemptsWithTimeout == null)
+            {
+                isRetryAttemptsWithTimeout = new[] { false };
+            }
+
+            var mockClock = new Mock<IClock>();
+
+            var nowSetup = mockClock.SetupSequence(c => c.UtcNow);
+            if (shouldNowBeAdded)
+            {
+                nowSetup.Returns(now);
+            }
+            foreach (var isTimeoutAttempt in isRetryAttemptsWithTimeout)
+            {
+                var passedTime = CalculateTime(isTimeoutAttempt);
+                nowSetup.Returns(now.AddSeconds(passedTime));
+            }
+            return mockClock;
+        }
+
+        private int CalculateTime(bool timeout)
         {
-            var fieldInfo = typeof(ClientSessionHandle).GetField("_disposed", BindingFlags.NonPublic | BindingFlags.Instance);
-            return (bool)fieldInfo.GetValue(obj);
+            return TransactionExecutorReflector.__transactionTimeout() + (timeout ? 10 : -10);
         }
 
-        public static IServerSession _serverSession(this ClientSessionHandle obj)
+        private void SetupTransactionState(ClientSessionHandle clientSession, bool isTransactionInProgress)
         {
-            var fieldInfo = typeof(ClientSessionHandle).GetField("_serverSession", BindingFlags.NonPublic | BindingFlags.Instance);
-            return (IServerSession)fieldInfo.GetValue(obj);
+            clientSession.WrappedCoreSession.CurrentTransaction.SetState(isTransactionInProgress ? CoreTransactionState.InProgress : CoreTransactionState.Aborted);
         }
     }
-}
+
+    internal static class ClientSessionHandleReflector
+    {
+        public static bool _disposed(this ClientSessionHandle obj) => (bool) Reflector.GetFieldValue(obj, nameof(_disposed));
+
+        public static IServerSession _serverSession(this ClientSessionHandle obj) => (IServerSession) Reflector.GetFieldValue(obj, nameof(_serverSession));
+    }
+
+    // TransactionExecutor
+    internal static class TransactionExecutorReflector
+    {
+        public static int __transactionTimeout() => (int) ((TimeSpan)Reflector.GetStaticFieldValue(typeof(TransactionExecutor), nameof(__transactionTimeout))).TotalSeconds;
+
+        public static bool CommitWithRetries(ClientSessionHandle session, DateTime startTime, IClock clock, CancellationToken cancellationToken)
+            => (bool)Reflector.InvokeStatic(typeof(TransactionExecutor), nameof(CommitWithRetries), (IClientSessionHandle)session, startTime, clock, cancellationToken);
+
+        public static Task<bool> CommitWithRetriesAsync(ClientSessionHandle session, DateTime startTime, IClock clock, CancellationToken cancellationToken)
+            => (Task<bool>)Reflector.InvokeStatic(typeof(TransactionExecutor), nameof(CommitWithRetriesAsync), (IClientSessionHandle)session, startTime, clock, cancellationToken);
```

Comment:
Instead of casting session parameter to IClientSessionHandle can we just declare the parameter type to be IClientSessionHandle?

I believe that should work. In general we want the parameter types in the reflectors to match exactly the parameter types in the methods being called via reflection.

Reply:
Done